wallet-new/src/Cardano/Wallet/WalletLayer/Kernel/Active.hs

```diff
@@ -61,6 +71,110 @@ estimateFees activeWallet grouping regulation payment = liftIO $ do
         withExceptT EstimateFeesError $ ExceptT $
           Kernel.estimateFees activeWallet opts accId payees
 
+-- | Creates a raw transaction.
+--
+-- NOTE: this function does /not/ perform a payment, it just creates a new
+-- transaction which will be signed and submitted to the blockchain later.
+createUnsignedTx :: MonadIO m
+                 => Kernel.ActiveWallet
+                 -> InputGrouping
+                 -> ExpenseRegulation
+                 -> V1.Payment
+                 -> m (Either NewTransactionError (Tx, NonEmpty (Address, [Word32])))
+createUnsignedTx activeWallet grouping regulation payment = liftIO $ do
+    let spendingPassword = maybe mempty coerce $ V1.pmtSpendingPassword payment
+        walletPassive = Kernel.walletPassive activeWallet
+    policy <- Node.getFeePolicy (Kernel.walletPassive activeWallet ^. Kernel.walletNode)
+    res <- runExceptT $ do
+        (opts, accId, payees) <- withExceptT NewTransactionWalletIdDecodingFailed $
+            setupPayment policy grouping regulation payment
+        (db, unsignedTx, utxo) <- withExceptT NewTransactionError $ ExceptT $
+            Kernel.newUnsignedTransaction activeWallet opts accId payees
+        return (db, unsignedTx, utxo, accId)
+    case res of
+        Left e -> return (Left e)
+        Right (db, unsignedTx, _utxo, srcAccId) -> do
+            -- We have to provide source addresses and derivation paths for this transaction.
+            -- It will be used by external party to provide a proof that it has a right to
+            -- spend this money.
+            let srcAddresses    = extractSrcAddressesFrom unsignedTx
+                srcHdAddresses  = NE.map (toHdAddress db) srcAddresses
+                derivationPaths = buildDerivationPathsFor srcHdAddresses srcAccId
+            -- Now we have to generate the change addresses needed,
+            -- because 'newUnsignedTransaction' function cannot do it by itself.
+            changeAddressesRes <- runExceptT $ withExceptT Kernel.NewTransactionErrorCreateAddressFailed $
+                genChangeOuts (utxChange unsignedTx) srcAccId spendingPassword walletPassive
+            case changeAddressesRes of
+                Left e -> return (Left $ NewTransactionError e)
+                Right changeAddresses -> do
+                    let allOutputs = (NE.toList $ utxOutputs unsignedTx) ++ changeAddresses
+                        unsignedTxWithChange = unsignedTx { utxOutputs = NE.fromList allOutputs }
+                        tx = toRegularTx unsignedTxWithChange
+                    return (Right (tx, NE.zip srcAddresses derivationPaths))
+  where
+    -- | Generates the list of change outputs from a list of change coins.
+    genChangeOuts :: MonadIO m
```

Comment:
The wallet layer seems to be doing _too much_ here (that's my gut feeling). If you compare the functions `pay` and `estimateFee` is the same module, you will see they are quite lean, they simply offload the hard work to the Kernel and they simply perform the wrapping/unwrapping of the data. I would suggest doing something similar to what we did there.

Note that once you start writing Kernel code, also your return types and the types you will be dealing with will change.